extensions/eclipse-che-theia-preferences-provider-extension/src/browser/che-tasks-preferences.ts

```diff
@@ -0,0 +1,138 @@
+import { che as cheApi } from '@eclipse-che/api';
+import { CheApiService } from '@eclipse-che/theia-plugin-ext/lib/common/che-protocol';
+import { TaskCustomization } from '@theia/task/lib/common';
+import { AsyncPreferenceProvider } from './async-preference-provider';
+import URI from '@theia/core/lib/common/uri';
+import * as jsoncparser from 'jsonc-parser';
+
+export const CHE_TASK_TYPE: string = 'che';
+export const MACHINE_NAME_ATTRIBUTE: string = 'machineName';
+export const PREVIEW_URL_ATTRIBUTE: string = 'previewUrl';
+export const WORKING_DIR_ATTRIBUTE: string = 'workingDir';
+export const COMPONENT_ALIAS_ATTRIBUTE: string = 'componentAlias';
+
+/**
+ *  This class injects commands form the workspace definition as tasks into theia preferences
+ */
+@injectable()
+export class CheTasksPreferences extends AsyncPreferenceProvider {
+
+    protected getScope(): PreferenceScope {
+        return PreferenceScope.User;
+    }
+
+    @inject(UserConfigsPreferenceProvider)
+    protected wrappedProvider: UserConfigsPreferenceProvider;
+
+    @inject(CheApiService)
+    protected cheApiService: CheApiService;
+
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    protected async load(): Promise<{ [key: string]: any; }> {
+        const originalPrefs = this.wrappedProvider.getPreferences();
+        const workspace: cheApi.workspace.Workspace = await this.cheApiService.currentWorkspace();
+        if (workspace && workspace.runtime && workspace.runtime.commands) {
+            const originalTasks: { tasks: TaskCustomization[] } = originalPrefs['tasks'] || {
+                tasks: []
+            };
+            originalTasks.tasks = originalTasks.tasks || [];
+            for (const command of workspace.runtime.commands) {
+                if (command.type === 'exec') {
+                    fromExecCommand(command, cmd => originalTasks.tasks.push(cmd));
+                } else if (command.type === 'vscode-task') {
+                    fromVSCodeTaskCommand(command, cmd => originalTasks.tasks.push(cmd));
+                }
+            };
+            return {
+                ...originalPrefs,
+                tasks: originalTasks
+            };
+
+        } else {
+            return originalPrefs;
+        }
+    }
+
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    doSetPreference(key: string, value: any, resourceUri?: string | undefined): Promise<boolean> {
+        return this.wrappedProvider.setPreference(key, value, resourceUri);
+    }
+
+    getConfigUri(): URI | undefined {
+        return this.wrappedProvider.getConfigUri();
+    }
+
+    getDomain(): string[] | undefined {
+        return this.wrappedProvider.getDomain();
+    }
+}
+
+function fromVSCodeTaskCommand(command: cheApi.workspace.Command, handler: (task: TaskCustomization) => void): void {
+    if (command && command.attributes && command.attributes.actionReferenceContent) {
```

Comment:
Sorry, I don't understand why `TaskCustomization` is used here. 
Can we use `TaskConfiguration` instead?

Reply:
Because TaskConfiguration have a mandatory scope, which vs code task customizations do not have.

Reply:
Generally, I think we should get rid of TaskConfiguration when talking about non-contributed things. I am not aware of an instance where it the differentiation makes sense.

Reply:
You can see how `TaskCustomization` is used on `theia` side.
For example, `task-service` [has](https://github.com/eclipse-theia/theia/blob/bb43e9ea7508687932e386a344f2afc0e915e451/packages/task/src/browser/task-service.ts#L885):
 `getTaskCustomization(task: TaskConfiguration): Promise<TaskCustomization>`

The method is used to get `TaskCustomization` based on given `TaskConfiguration`:
- at running a task we have `TaskConfiguration` for running and [get](https://github.com/eclipse-theia/theia/blob/bb43e9ea7508687932e386a344f2afc0e915e451/packages/task/src/browser/task-service.ts#L488)  `customizationObject` to resolve `problemMatchers` for example

- [it's impossible](https://github.com/eclipse-theia/theia/blob/bb43e9ea7508687932e386a344f2afc0e915e451/packages/task/src/browser/task-configurations.ts#L250) to get `TaskCustomization` if it's **NOT** `detected` task

So I don't understand how we can consider a `che` task as `TaskCustomization`. Customization for what?

Reply:
Yes, but TaskCustomization and TaskConfiguration are interfaces: they are simply promises that an object has certain properties of certain types: so every TaskConfiguration is also a TaskCustomization. So the object we're returning here does not have the mandatory "scope" of the appropriate type. So it does not conform to the TaskConfiguration interface.

> it's impossible to get TaskCustomization if it's NOT detected task

I don't understand what you're trying to say here: could you explain differently?

In general, I don't think it ever makes sense to talk about a TaskConfiguration for anything that is not a detected task.

Reply:
Do you consider `che` task as `detected` or as `configured` task?

Reply:
The way it is implemented now they are configured tasks (aka TaskCustomizations). Maybe that _is_ the wrong approach and we should present them as detected tasks. At this point, we could call "configure" on the che tasks and would be able to set stuff like "problemMatcher" in the user-level "tasks.json". What do you think.